src/main/java/io/cryostat/core/reports/InterruptibleReportGenerator.java

```diff
@@ -0,0 +1,350 @@
+                                        .stream()
+                                        .map(executor::submit)
+                                        .collect(Collectors.toList()));
+                        Collection<Result> results = new HashSet<>();
+                        for (Future<Result> future : resultFutures) {
+                            results.add(future.get());
+                        }
+
+                        List<HtmlResultGroup> groups = loadResultGroups();
+
+                        return transform(
+                                RulesHtmlToolkit.generateStructuredHtml(
+                                        new SimpleResultProvider(results, groups),
+                                        groups,
+                                        new HashMap<String, Boolean>(),
+                                        true));
+                    } catch (InterruptedException
+                            | IOException
+                            | ExecutionException
+                            | CouldNotLoadRecordingException e) {
+                        resultFutures.forEach(
+                                f -> {
+                                    if (!f.isDone()) {
+                                        f.cancel(true);
+                                    }
+                                });
+                        logger.warn(e);
+                        return "<html>"
+                                + " <head></head>"
+                                + " <body>"
+                                + "  <div>"
+                                + e.getMessage()
+                                + "  </div>"
+                                + " </body>"
+                                + "</html>";
+                    }
+                });
+    }
+
+    String transform(String report) {
+        if (transformers.isEmpty()) {
+            return report;
+        }
+        try {
+            org.jsoup.nodes.Document document = Jsoup.parse(report);
+            transformers.forEach(
+                    t -> {
+                        document.select(t.selector())
+                                .forEach(
+                                        el -> {
+                                            el.html(t.innerHtml(el.html()));
+                                            t.attributes()
+                                                    .entrySet()
+                                                    .forEach(
+                                                            e -> el.attr(e.getKey(), e.getValue()));
+                                        });
+                    });
+            return document.outerHtml();
+        } catch (Exception e) {
+            logger.warn(e);
+            return report;
+        }
+    }
+
+    /**
+     * Everything below this point is forked from JMC's JfrHtmlRulesReport and RulesToolkit classes,
```

Comment:
Out of curiosity, how would you know whether or not to update cryostat-core if jmc makes changes to these method implementations?

Reply:
It's totally manual and best-effort at this point. If this continues to compile and our integration tests pass then we can feel pretty safe that nothing was fundamentally broken, but if there are any behavioural changes under the hood in JMC, that's just up to me/maintainers to review the changes and adapt our -core as needed. So if the JMC classes we're relying on become interruptible on their own to begin with, or they refactor their other existing cancellable/interruptible implementation to not be tightly coupled to Eclipse, then we could just drop my implementation here for example.

Reply:
Do you think any of these changes make sense to integrate into upstream JMC in the future?

Reply:
Not sure. AFAICT this entrypoint into the rules processing system isn't really used by JMC anymore - they appear to just be using the Eclipse-coupled entrypoint instead. I'm not sure of the history between those two, or if I'm missing a place where the entrypoint we use is actually still used by JMC somewhere.

If it is valuable to JMC at all then I would certainly be happy to upstream the change.